### PR TITLE
INT-1000 - compatibility update  for submission

### PR DESCRIPTION
**Pull request by Francois-Gomis**

### CODE REVIEW

File: Helpers/Logger.php [hunk 1]
```diff
@@ -54,26 +53,27 @@ public function __construct(
...
     /**
      * @param int $level
      * @param string $message
      * @param array $context
+     * @param DateTimeImmutable|null $dateTimeImmutable
+     *
      * @return bool
      */
-    public function addRecord($level, $message, array $context = []): bool
+    public function addRecord($level, $message, array $context = [], ?DateTimeImmutable $dateTimeImmutable = null): bool
```
[the remark this answers is not in the document]
Comment (syjust-alma) [reply Helpers/Logger.php:76]: or it only have to depends on a recent version of Monolog which will works with older Magento releases ?

File: Setup/Patch/Data/UpdateCancelUrl.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+use Alma\MonthlyPayments\Gateway\Config\Config;
+use Alma\MonthlyPayments\Helpers\Logger;
+use Alma\MonthlyPayments\Model\Adminhtml\Config\ApiUrl\CustomerCancelUrl;
+use Exception;
+use Magento\Framework\App\ResourceConnection;
+use Magento\Framework\Exception\NoSuchEntityException;
+use Magento\Framework\Setup\Patch\DataPatchInterface;
+use Magento\Payment\Gateway\Config\Config as ConfigGateway;
+
+class UpdateCancelUrl implements DataPatchInterface
```
Comment (ClaireBonacheraAlma) [inline Setup/Patch/Data/UpdateCancelUrl.php:14]: missing class doccomment

File: Setup/Patch/Data/UpdateCancelUrl.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    public function apply(): void
+    {
+        $path = sprintf(ConfigGateway::DEFAULT_PATH_PATTERN, Config::CODE, Config::CONFIG_CUSTOMER_CANCEL_URL);
+        $oldProcessedPath = $this->customerCancelUrl->getOldDefaultUrl();
+        $newProcessedPath = $this->customerCancelUrl->processValue('');
+        $tableName = $this->getCoreTable();
+        $replace = ['value' => $newProcessedPath];
+        $where = [self::PATH_EQUAL => $path, 'value = ?' => $oldProcessedPath];
+        try {
+            $this->resourceConnection->getConnection()->update($tableName, $replace, $where);
```
Comment (ClaireBonacheraAlma) [inline Setup/Patch/Data/UpdateCancelUrl.php:62]: you can put all the code in the try

File: Setup/Patch/Data/UpdateMerchantIdPath.php [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+use Alma\MonthlyPayments\Helpers\AlmaClient;
+use Alma\MonthlyPayments\Helpers\ConfigHelper;
+use Alma\MonthlyPayments\Helpers\Exceptions\AlmaClientException;
+use Alma\MonthlyPayments\Helpers\Logger;
+use Magento\Framework\App\Config\ScopeConfigInterface;
+use Magento\Framework\App\ResourceConnection;
+use Magento\Framework\Setup\Patch\DataPatchInterface;
+use Magento\Payment\Gateway\Config\Config as ConfigGateway;
+
+class UpdateMerchantIdPath implements DataPatchInterface
```
Comment (ClaireBonacheraAlma) [inline Setup/Patch/Data/UpdateMerchantIdPath.php:17]: doccomment

File: Setup/Patch/Data/UpdateMerchantIdPath.php [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        $apiQuery = $this->resourceConnection->getConnection()->select()
+            ->from($tableName, ['config_id','scope','scope_id', 'path', 'value'])
+            ->where(self::PATH_EQUAL, 'payment/alma_monthly_payments/live_api_key')
+            ->orWhere(self::PATH_EQUAL, 'payment/alma_monthly_payments/test_api_key');
+
+        $apiKeysRows = $this->resourceConnection->getConnection()->fetchAll($apiQuery);
+
+        foreach ($apiKeysRows as $apiKeysRow) {
+            preg_match('/(live|test)/', $apiKeysRow['path'], $matches);
+            $apiKey = $this->scopeConfig->getValue($apiKeysRow['path'], $apiKeysRow['scope'], $apiKeysRow['scope_id']);
```
Comment (ClaireBonacheraAlma) [inline Setup/Patch/Data/UpdateMerchantIdPath.php:94]: for lisibility and unicity, I will have move the code inside the forech to an other function

Comment (Francois-Gomis) [reply Setup/Patch/Data/UpdateMerchantIdPath.php:94]: done

File: Setup/Patch/Data/UpdateMerchantIdPath.php (hunk 4 above)
Comment (Francois-Gomis) [reply Setup/Patch/Data/UpdateMerchantIdPath.php:17]: done

**Review by ClaireBonacheraAlma — APPROVED** [review]

File: Setup/Patch/Data/UpdateCancelUrl.php (hunk 3 above)
Comment (syjust-alma) [reply Setup/Patch/Data/UpdateCancelUrl.php:65]: not always true ;)

**Review by syjust-alma — APPROVED** [review]

**Pull request merged**